### PR TITLE
Segmentation: handle former donor

### PR DESCRIPTION
**Pull request by adekbadek**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds "former donor" segmentation handling, and receiving data about a cancelled recurring donation.

### How to test the changes in this Pull Request:

_Follow testing steps in https://github.com/Automattic/newspack-plugin/pull/1966_

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by miguelpeixe — CHANGES_REQUESTED** [review]
I successfully created a segment and a prompt for "Former donors". Canceling immediately through the Stripe dashboard triggers the hook and the `donation_cancelled` event is stored to `wp_newspack_campaigns_reader_events` but the prompt never shows up. I did receive a prompt targeted to "Donors" after canceling the subscription.

File: api/campaigns/class-campaign-data-utils.php [hunk 1]
```diff
@@ -340,4 +368,18 @@ public static function is_inline( $popup ) {
 	public static function is_above_header( $popup ) {
 		return isset( $popup->t ) && 'a' === $popup->t;
 	}
+
+	/**
+	 * Get all events types.
+	 */
+	public static function get_all_events_types() {
```
Comment (dkoo) [inline api/campaigns/class-campaign-data-utils.php:375]: "all event types" isn't entirely accurate here, as there are also `prompt_seen` and `prompt_dismissed` types. Maybe `get_all_reader_events_types`?

**Review by dkoo — CHANGES_REQUESTED** [review]
One reason this can easily fail is if the cancellation happens in a separate session (with a different client ID) from the original donation. This is probably more likely to happen when cancelling a subscription since it might happen weeks or months later after the first donation. #901 would fix that issue.

Following the testing steps by donating in a new session, cancelling the subscription in Stripe and then continuing to navigate the site in the same session, I see that this is working.

One nit below about a method name from me, but I'd also await confirmation from @miguelpeixe on whether it works for him in the same session as well.

File: api/campaigns/class-campaign-data-utils.php (hunk 1 above)
Comment (adekbadek) [reply api/campaigns/class-campaign-data-utils.php]: Refactored to `get_reader_events_types` in f379d0b803eefd128911c0bcbb533ec819427320

**Comment by adekbadek** [comment]
> One reason this can easily fail is if the cancellation happens in a separate session (with a different client ID) from the original donation. 

Session is not a factor here – the client ID is a metadata on the Stripe customer. The session during which the cancellation happens does not matter at all.

**Comment by dkoo** [comment]
Session does matter somewhat because until we solve #901 (slated for RAS phase 2 in the near future), we don't yet have a means to tie different client IDs together even if they're tied to the same email. This is the same issue as not knowing about past subscriptions/donations if you log into the same account with a different device or in a different browser session, I'm just highlighting that the particular use case of cancelling recurring donations is more likely to encounter this issue because a cancellation is likely happening weeks or months after the initial donation.

**Comment by adekbadek** [comment]
I think we were talking about two different things :) By "this can easily fail" do you mean a following scenario:

1. reader donates on device A
2. they log in on device B
3. they cancel the subscription – now they're recognised as a former donor only on device B

?

**Review by dkoo — APPROVED** [review]
That's the scenario I was talking about, yes. It's not just different devices, but different client IDs—expected given our single client ID logic, but I think it's more likely to occur with cancellations because browser sessions/client IDs may not be likely to persist for months between donation and cancellation.

Approving because it works with the current logic, but I'd also like @miguelpeixe to confirm if it works for him or if there's another scenario that I didn't encounter.

**Comment by miguelpeixe** [comment]
@dkoo did you successfully render a prompt segmented to former donors? I'm still having the exact same issue [reported before](https://github.com/Automattic/newspack-popups/pull/954#pullrequestreview-1095381630):

> I successfully created a segment and a prompt for "Former donors". Canceling immediately through the Stripe dashboard triggers the hook and the `donation_cancelled` event is stored to `wp_newspack_campaigns_reader_events` but the prompt never shows up. I did receive a prompt targeted to "Donors" after canceling the subscription.

**Comment by dkoo** [comment]
> @dkoo did you successfully render a prompt segmented to former donors?

I did, which is why I approved it. The "former donors" prompt only appears in the same session as the original donation.

**Comment by miguelpeixe** [comment]
I was in the same session 🤔

**Review by miguelpeixe — APPROVED** [review]

**Pull request merged**